web/pdf_viewer.js

```diff
@@ -953,7 +929,176 @@ class PDFViewer {
+    let firstPage = visible.first;
+    let stillFullyVisible = false;
+
+    for (let i = 0, ii = visiblePages.length; i < ii; ++i) {
+      let page = visiblePages[i];
+
+      if (page.percent < 100) {
+        break;
+      }
+      if (page.id === currentId) {
+        stillFullyVisible = true;
+        break;
+      }
+    }
+
+    if (!stillFullyVisible) {
+      currentId = visiblePages[0].id;
+    }
+
+    if (!this.isInPresentationMode) {
+      this._setCurrentPageNumber(currentId);
+    }
+
+    this._updateLocation(firstPage);
+    this.eventBus.dispatch('updateviewarea', {
+      source: this,
+      location: this._location,
+    });
+  }
+}
+
+class PDFSinglePageViewer extends BaseViewer {
+  constructor(options) {
+    super(options);
+
+    this.eventBus.on('pagesinit', (evt) => {
+      // Since the pages are placed in a `DocumentFragment`, make sure that
+      // the current page becomes visible upon loading of the document.
+      this._ensurePageViewVisible();
+    });
+  }
+
+  _resetView() {
+    super._resetView();
+    this._previousPageNumber = 1;
+    this._shadowViewer = document.createDocumentFragment();
+  }
+
+  /**
+   * @returns {boolean} scrolledDown
+   */
+  _ensurePageViewVisible() {
+    let pageView = this._pages[this._currentPageNumber - 1];
+    let previousPageView = this._pages[this._previousPageNumber - 1];
+
+    let viewerNodes = this.viewer.childNodes;
+    switch (viewerNodes.length) {
+      case 0: // Should *only* occur on initial loading.
+        this.viewer.appendChild(pageView.div);
+        break;
+      case 1: // The normal page-switching case.
+        if (viewerNodes[0] !== previousPageView.div) {
+          throw new Error(
+            '_ensurePageViewVisible: unexpected previously visible page.');
+        }
+        if (pageView === previousPageView) {
+          break; // The correct page is already visible.
+        }
+        // Switch visible pages, and ensure that rendering is triggered.
+        this._shadowViewer.appendChild(previousPageView.div);
+        this.viewer.appendChild(pageView.div);
+
+        this.update();
+        break;
+      default:
+        throw new Error(
+          '_ensurePageViewVisible: Only one page should be visible at a time.');
+    }
+    // Keep track of the actual "scroll" direction...
+    let scrolledDown = this._currentPageNumber >= this._previousPageNumber;
+    // ... before updating the previous pageNumber.
+    this._previousPageNumber = this._currentPageNumber;
+
+    return scrolledDown;
+  }
+
+  _scrollIntoView({ pageDiv, pageSpot = null, pageNumber = null, }) {
+    if (pageNumber) {
+      // Ensure that `this._currentPageNumber` is always correct.
+      this._setCurrentPageNumber(pageNumber);
+    }
+    let scrolledDown = this._ensurePageViewVisible();
+
+    scrollIntoView(pageDiv, pageSpot);
+
+    // Always update the scroll direction, since it most likely isn't correct.
+    this.scroll.down = scrolledDown;
```

Comment:
Since the code currently relies on `this.scroll` when determining which page should be pre-rendered, and I didn't feel like refactoring all of that here too, this patch is still updating `this.scroll.down` according to the page switching direction.
However it's now done in such a way that it at least should be consistent, since it waits for the next scroll event; and unless someone complains violently I'd lean towards keeping this for now.